zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,911 @@
+        move |input| {
+            map_res(
+                separated_pair(indexed_name, char('='), recognize(qchars)),
+                move |r| to_indexed_param(params, r),
+            )(input)
+        }
+    }
+
+    /// Extension for the `alphanumeric0` parser which extends that parser with
+    /// the permitted characters identified in
+    fn alphanum_or(allowed: String) -> impl (Fn(&str) -> IResult<&str, &str>) {
+        move |input| {
+            input.split_at_position_complete(|item| {
+                let c = item.as_char();
+                !(c.is_alphanum() || allowed.contains(c))
+            })
+        }
+    }
+
+    pub fn qchars(input: &str) -> IResult<&str, &str> {
+        alphanum_or("-._~!$'()*+,;:@%".to_string())(input)
+    }
+
+    pub fn namechars(input: &str) -> IResult<&str, &str> {
+        alphanum_or("+-".to_string())(input)
+    }
+
+    pub fn indexed_name<'a>(input: &'a str) -> IResult<&'a str, (&'a str, Option<&'a str>)> {
+        let paramname = recognize(tuple((alpha1, namechars)));
+
+        tuple((
+            paramname,
+            opt(preceded(
+                char('.'),
+                recognize(tuple((one_of("123456789"), digit0))),
+            )),
+        ))(input)
+    }
+
+    pub fn parse_amount<'a>(input: &'a str) -> IResult<&'a str, Amount> {
+        map_res(
+            tuple((digit1, opt(preceded(char('.'), digit1)))),
+            |(whole_s, decimal_s): (&str, Option<&str>)| {
+                let coins: i64 = whole_s
+                    .to_string()
+                    .parse::<i64>()
+                    .map_err(|e| e.to_string())?;
+
+                let zats: i64 = match decimal_s {
+                    Some(d) => d
+                        .char_indices()
+                        .map(|(i, c)| {
+                            let digit: i64 = c.to_digit(10).unwrap().into();
+                            digit * 10i64.pow((7 - i).try_into().unwrap())
+                        })
+                        .sum::<i64>()
+                        .into(),
+                    None => 0,
+                };
+
+                let amt = coins * COIN + zats;
+
+                Amount::from_nonnegative_i64(amt)
+                    .map_err(|_| format!("Not a valid zat amount: {}", amt))
+            },
+        )(input)
+    }
+
+    fn to_indexed_param<'a, P: consensus::Parameters>(
+        params: &'a P,
+        ((name, iopt), value): ((&str, Option<&str>), &str),
+    ) -> Result<IndexedParam, String> {
+        let param = match name {
+            "address" => Address::parse(params, value)
+                .map_err(|e| match e {
+                    AddrParseError::TAddrError(_) => format!(
+                        "Could not interpret {} as a valid transparent address.",
+                        value
+                    ),
+                    AddrParseError::SAddrError(_) => {
+                        format!("Could not interpret {} as a valid sapling address.", value)
```

Comment:
```suggestion
                        format!("Could not interpret {} as a valid Sapling address.", value)
```

Reply:
Should we produce a more specific error for something that looks like a Sprout address (by prefix)?